File: scala/scalasearch/src/main/scala/scalasearch/SearchSettings.scala
package scalasearch

import scalafind.Color.Color
import scalafind.FileType.FileType
import scalafind.SortBy.SortBy
import scalafind.{Color, DefaultFindSettings, FindSettings, SortBy}

import java.nio.file.Path
import java.time.LocalDateTime
import scala.util.matching.Regex

object DefaultSearchSettings {
  val archivesOnly: Boolean = DefaultFindSettings.archivesOnly
  val colorize: Boolean = DefaultFindSettings.colorize
  val debug: Boolean = DefaultFindSettings.debug
  val dirColor: Color = DefaultFindSettings.dirColor
  val extColor: Color = DefaultFindSettings.extColor
  val fileColor: Color = DefaultFindSettings.fileColor
  val firstMatch = false
  val followSymlinks: Boolean = DefaultFindSettings.followSymlinks
  val includeHidden: Boolean = DefaultFindSettings.includeHidden
  val lineColor: Color = Color.GREEN
  val linesAfter = 0
  val linesBefore = 0
  val maxDepth: Int = DefaultFindSettings.maxDepth
  val maxSize: Long = DefaultFindSettings.maxSize
  val minDepth: Int = DefaultFindSettings.minDepth
  val minSize: Long = DefaultFindSettings.minSize
  val maxLineLength = 150
  var multiLineSearch = false
  val printDirs: Boolean = DefaultFindSettings.printDirs
  val printFiles: Boolean = DefaultFindSettings.printFiles
  val printLines = false
  val printMatches = false
  var printResults = false
  val printUsage: Boolean = DefaultFindSettings.printUsage
  val printVersion: Boolean = DefaultFindSettings.printVersion
  val recursive: Boolean = DefaultFindSettings.recursive
  var searchArchives = false
  var textFileEncoding: String = "UTF-8"
  var uniqueLines = false
  val verbose: Boolean = DefaultFindSettings.verbose
}

case class SearchSettings(archivesOnly: Boolean = DefaultSearchSettings.archivesOnly,
                          colorize: Boolean = DefaultSearchSettings.colorize,
                          debug: Boolean = DefaultSearchSettings.debug,
                          dirColor: Color = DefaultSearchSettings.dirColor,
                          extColor: Color = DefaultSearchSettings.extColor,
                          fileColor: Color = DefaultSearchSettings.fileColor,
                          firstMatch: Boolean = DefaultSearchSettings.firstMatch,
                          followSymlinks: Boolean = DefaultSearchSettings.followSymlinks,
                          inArchiveExtensions: Set[String] = Set.empty[String],
                          inArchiveFilePatterns: Set[Regex] = Set.empty[Regex],
                          includeHidden: Boolean = DefaultSearchSettings.includeHidden,
                          inDirPatterns: Set[Regex] = Set.empty[Regex],
                          inExtensions: Set[String] = Set.empty[String],
                          inFilePatterns: Set[Regex] = Set.empty[Regex],
                          inFileTypes: Set[FileType] = Set.empty[FileType],
                          inLinesAfterPatterns: Set[Regex] = Set.empty[Regex],
                          inLinesBeforePatterns: Set[Regex] = Set.empty[Regex],
                          lineColor: Color = DefaultSearchSettings.lineColor,
                          linesAfter: Int = DefaultSearchSettings.linesAfter,
                          linesAfterToPatterns: Set[Regex] = Set.empty[Regex],
                          linesAfterUntilPatterns: Set[Regex] = Set.empty[Regex],
                          linesBefore: Int = DefaultSearchSettings.linesBefore,
                          maxDepth: Int = DefaultSearchSettings.maxDepth,
                          maxLastMod: Option[LocalDateTime] = None,
                          maxLineLength: Int = DefaultSearchSettings.maxLineLength,
                          maxSize: Long = DefaultSearchSettings.maxSize,
                          minDepth: Int = DefaultSearchSettings.minDepth,
                          minLastMod: Option[LocalDateTime] = None,
                          minSize: Long = DefaultSearchSettings.minSize,
                          multiLineSearch: Boolean = DefaultSearchSettings.multiLineSearch,
                          outArchiveExtensions: Set[String] = Set.empty[String],
                          outArchiveFilePatterns: Set[Regex] = Set.empty[Regex],
                          outDirPatterns: Set[Regex] = Set.empty[Regex],
                          outExtensions: Set[String] = Set.empty[String],
                          outFilePatterns: Set[Regex] = Set.empty[Regex],
                          outFileTypes: Set[FileType] = Set.empty[FileType],
                          outLinesAfterPatterns: Set[Regex] = Set.empty[Regex],
                          outLinesBeforePatterns: Set[Regex] = Set.empty[Regex],
                          paths: Set[Path] = Set.empty[Path],
                          printDirs: Boolean = DefaultSearchSettings.printDirs,
                          printFiles: Boolean = DefaultSearchSettings.printFiles,
                          printLines: Boolean = DefaultSearchSettings.printLines,
                          printMatches: Boolean = DefaultSearchSettings.printMatches,
                          printResults: Boolean = DefaultSearchSettings.printResults,
                          printUsage: Boolean = DefaultSearchSettings.printUsage,
                          printVersion: Boolean = DefaultSearchSettings.printVersion,
                          recursive: Boolean = DefaultSearchSettings.recursive,
                          var searchArchives: Boolean = DefaultSearchSettings.searchArchives,
                          searchPatterns: Set[Regex] = Set.empty[Regex],
                          sortBy: SortBy = SortBy.FilePath,
                          textFileEncoding: String = DefaultSearchSettings.textFileEncoding,
                          uniqueLines: Boolean = DefaultSearchSettings.uniqueLines,
                          sortCaseInsensitive: Boolean = false,
                          sortDescending: Boolean = false,
                          var verbose: Boolean = DefaultSearchSettings.verbose) {

  searchArchives = archivesOnly || searchArchives
  verbose = debug || verbose

  val findSettings: FindSettings = FindSettings(
    archivesOnly = archivesOnly,
    colorize = colorize,
    debug = debug,
    dirColor = dirColor,
    extColor = extColor,
    fileColor = fileColor,
    followSymlinks = followSymlinks,
    inArchiveExtensions = inArchiveExtensions,
    inArchiveFilePatterns = inArchiveFilePatterns,
    includeHidden = includeHidden,
    inDirPatterns = inDirPatterns,
    inExtensions = inExtensions,
    inFilePatterns = inFilePatterns,
    inFileTypes = inFileTypes,
    includeArchives = searchArchives,
    maxDepth = maxDepth,
    maxLastMod = maxLastMod,
    maxSize = maxSize,
    minDepth = minDepth,
    minLastMod = minLastMod,
    minSize = minSize,
    outArchiveExtensions = outArchiveExtensions,
    outArchiveFilePatterns = outArchiveFilePatterns,
    outDirPatterns = outDirPatterns,
    outExtensions = outExtensions,
    outFilePatterns = outFilePatterns,
    outFileTypes = outFileTypes,
    paths = paths,
    printDirs = printDirs,
    printFiles = printFiles,
    printUsage = printUsage,
    printVersion = printVersion,
    recursive = recursive,
    sortBy = sortBy,
    sortCaseInsensitive = sortCaseInsensitive,
    sortDescending = sortDescending,
    verbose = verbose)

  def addExtensions(exts: String, extensions: Set[String]): Set[String] = {
    findSettings.addExtensions(exts, extensions)
  }

  def getLastModFromString(lastModString: String): Option[LocalDateTime] = {
    findSettings.getLastModFromString(lastModString)
  }

  def hasLinesBefore: Boolean = {
    linesBefore > 0 || hasLinesBeforePatterns
  }

  def hasLinesBeforePatterns: Boolean = {
    inLinesBeforePatterns.nonEmpty || outLinesBeforePatterns.nonEmpty
  }

  def hasLinesAfter: Boolean = {
    linesAfter > 0 || hasLinesAfterPatterns
  }

  def hasLinesAfterPatterns: Boolean = {
    inLinesAfterPatterns.nonEmpty || outLinesAfterPatterns.nonEmpty
  }

  def hasLinesAfterToPatterns: Boolean = linesAfterToPatterns.nonEmpty

  def hasLinesAfterUntilPatterns: Boolean = linesAfterUntilPatterns.nonEmpty

  def hasLinesAfterToOrUntilPatterns: Boolean = {
    hasLinesAfterToPatterns || hasLinesAfterUntilPatterns
  }

  override def toString: String = {
    "SearchSettings(" +
      "archivesOnly: " + archivesOnly +
      ", colorize: " + colorize +
      ", debug: " + debug +
      ", firstMatch: " + firstMatch +
      ", followSymlinks: " + followSymlinks +
      ", inArchiveExtensions: " + inArchiveExtensions +
      ", inArchiveFilePatterns: " + inArchiveFilePatterns +
      ", includeHidden: " + includeHidden +
      ", inDirPatterns: " + inDirPatterns +
      ", inExtensions: " + inExtensions +
      ", inFilePatterns: " + inFilePatterns +
      ", inFileTypes: " + inFileTypes +
      ", linesAfter: " + linesAfter +
      ", linesBefore: " + linesBefore +
      ", maxDepth: " + maxDepth +
      ", maxLastMod: " + maxLastMod +
      ", maxLineLength: " + maxLineLength +
      ", maxSize: " + maxSize +
      ", minDepth: " + minDepth +
      ", minLastMod: " + minLastMod +
      ", minSize: " + minSize +
      ", multiLineSearch: " + multiLineSearch +
      ", outArchiveExtensions: " + outArchiveExtensions +
      ", outArchiveFilePatterns: " + outArchiveFilePatterns +
      ", outDirPatterns: " + outDirPatterns +
      ", outExtensions: " + outExtensions +
      ", outFilePatterns: " + outFilePatterns +
      ", outFileTypes: " + outFileTypes +
      ", paths: " + paths  +
      ", printDirs: " + printDirs +
      ", printFiles: " + printFiles +
      ", printLines: " + printLines +
      ", printMatches: " + printMatches +
      ", printResults: " + printResults +
      ", printUsage: " + printUsage +
      ", printVersion: " + printVersion +
      ", recursive: " + recursive +
      ", searchArchives: " + searchArchives +
      ", searchPatterns: " + searchPatterns +
      ", sortBy: " + sortBy +
      ", sortCaseInsensitive: " + sortCaseInsensitive +
      ", sortDescending: " + sortDescending +
      ", textFileEncoding: " + textFileEncoding +
      ", uniqueLines: " + uniqueLines +
      ", verbose: " + verbose +
      ")"
  }
}

File: kotlin/ktsearch/src/main/kotlin/ktsearch/SearchSettings.kt
package ktsearch

import ktfind.FileType
import java.time.LocalDateTime

import ktfind.FindSettings
import ktfind.SortBy
import ktfind.fileTypeFromName

/**
 * @author cary on 7/23/16.
 */
data class SearchSettings(val archivesOnly: Boolean,
                          val colorize: Boolean,
                          val debug: Boolean,
                          val excludeHidden: Boolean,
                          val firstMatch: Boolean,
                          val inArchiveExtensions: Set<String>,
                          val inArchiveFilePatterns: Set<Regex>,
                          val inDirPatterns: Set<Regex>,
                          val inExtensions: Set<String>,
                          val inFilePatterns: Set<Regex>,
                          val inFileTypes: Set<FileType>,
                          val inLinesAfterPatterns: Set<Regex>,
                          val inLinesBeforePatterns: Set<Regex>,
                          val linesAfter: Int,
                          val linesAfterToPatterns: Set<Regex>,
                          val linesAfterUntilPatterns: Set<Regex>,
                          val linesBefore: Int,
                          val listDirs: Boolean,
                          val listFiles: Boolean,
                          val listLines: Boolean,
                          val maxLastMod: LocalDateTime?,
                          val maxLineLength: Int,
                          val maxSize: Int,
                          val minLastMod: LocalDateTime?,
                          val minSize: Int,
                          val multiLineSearch: Boolean,
                          val outArchiveExtensions: Set<String>,
                          val outArchiveFilePatterns: Set<Regex>,
                          val outDirPatterns: Set<Regex>,
                          val outExtensions: Set<String>,
                          val outFilePatterns: Set<Regex>,
                          val outFileTypes: Set<FileType>,
                          val outLinesAfterPatterns: Set<Regex>,
                          val outLinesBeforePatterns: Set<Regex>,
                          val paths: Set<String>,
                          val printResults: Boolean,
                          val printUsage: Boolean,
                          val printVersion: Boolean,
                          val recursive: Boolean,
                          val searchArchives: Boolean,
                          val searchPatterns: Set<Regex>,
                          val sortBy: SortBy,
                          val sortCaseInsensitive: Boolean,
                          val sortDescending: Boolean,
                          val textFileEncoding: String,
                          val uniqueLines: Boolean,
                          val verbose: Boolean) {
    val findSettings: FindSettings = FindSettings(archivesOnly,
        debug,
        excludeHidden,
        inArchiveExtensions,
        inArchiveFilePatterns,
        inDirPatterns,
        inExtensions,
        inFilePatterns,
        inFileTypes,
        searchArchives,
        listDirs,
        listFiles,
        maxLastMod,
        maxSize,
        minLastMod,
        minSize,
        outArchiveExtensions,
        outArchiveFilePatterns,
        outDirPatterns,
        outExtensions,
        outFilePatterns,
        outFileTypes,
        paths,
        printUsage,
        printVersion,
        recursive,
        sortBy,
        sortCaseInsensitive,
        sortDescending,
        verbose
    )
}

fun getDefaultSettings() : SearchSettings {
    return SearchSettings(
        archivesOnly = false,
        colorize = true,
        debug = false,
        excludeHidden = true,
        firstMatch = false,
        inArchiveExtensions = setOf(),
        inArchiveFilePatterns = setOf(),
        inDirPatterns = setOf(),
        inExtensions = setOf(),
        inFilePatterns = setOf(),
        inFileTypes = setOf(),
        inLinesAfterPatterns = setOf(),
        inLinesBeforePatterns = setOf(),
        linesAfter = 0,
        linesAfterToPatterns = setOf(),
        linesAfterUntilPatterns = setOf(),
        linesBefore = 0,
        listDirs = false,
        listFiles = false,
        listLines = false,
        maxLastMod = null,
        maxLineLength = 150,
        maxSize = 0,
        minLastMod = null,
        minSize = 0,
        multiLineSearch = false,
        outArchiveExtensions = setOf(),
        outArchiveFilePatterns = setOf(),
        outDirPatterns = setOf(),
        outExtensions = setOf(),
        outFilePatterns = setOf(),
        outFileTypes = setOf(),
        outLinesAfterPatterns = setOf(),
        outLinesBeforePatterns = setOf(),
        paths = setOf(),
        printResults = false,
        printUsage = false,
        printVersion = false,
        recursive = true,
        searchArchives = false,
        searchPatterns = setOf(),
        sortBy = SortBy.FILEPATH,
        sortCaseInsensitive = false,
        sortDescending = false,
        textFileEncoding = "UTF-8",
        uniqueLines = false,
        verbose = false)
}

fun addExtensions(ext: String, extensions: Set<String>): Set<String> {
    val exts = ext.split(',').filter { it.isNotEmpty() }
    return extensions.plus(exts)
}

fun addFileTypes(ft: String, filetypes: Set<FileType>): Set<FileType> {
    val fts = ft.split(',').filter { it.isNotEmpty() }.map { fileTypeFromName(it) }
    return filetypes.plus(fts)
}

fun setArchivesOnly(ss: SearchSettings, archivesOnly: Boolean): SearchSettings {
    return ss.copy(archivesOnly = archivesOnly, searchArchives = archivesOnly || ss.searchArchives)
}

fun setDebug(ss: SearchSettings, debug: Boolean): SearchSettings {
    return ss.copy(debug = debug, verbose = debug || ss.verbose)
}
